protocol/xmrmaker/claim.go

```diff
@@ -99,6 +108,53 @@ func (s *swapState) claimFunds() (ethcommon.Hash, error) {
 	return txHash, nil
 }
 
+// discoverRelayersAndClaim discovers available relayers on the network,
+func (s *swapState) discoverRelayersAndClaim() (ethcommon.Hash, error) {
+	relayers, err := s.Backend.DiscoverRelayers()
+	if err != nil {
+		return ethcommon.Hash{}, err
+	}
+
+	forwarderAddress, err := s.Contract().TrustedForwarder(&bind.CallOpts{})
+	if err != nil {
+		return ethcommon.Hash{}, err
+	}
+
+	calldata, err := getClaimTxCalldata(common.DefaultRelayerCommission, &s.contractSwap, s.getSecret())
+	if err != nil {
+		return ethcommon.Hash{}, err
+	}
+
+	req, err := relayer.CreateSubmitTransactionRequest(
+		s.ETHClient().PrivateKey(),
+		s.ETHClient().Raw(),
+		forwarderAddress,
+		s.contractAddr,
+		calldata,
+	)
+	if err != nil {
+		return ethcommon.Hash{}, err
+	}
+
+	for _, relayer := range relayers {
+		resp, err := s.Backend.SubmitTransactionToRelayer(relayer, req)
+		if err != nil {
+			log.Warnf("failed to submit tx to relayer with peer ID %s, trying next relayer: err: %s", relayer, err)
+			continue
+		}
+
+		err = waitAndCheck(s.ctx, s.ETHClient().Raw(), resp.TxHash)
```

Comment:
Is it expected that the relayer waited for at least one confirmation before the request returns? How long do we wait for a receipt? We may need to give up much sooner than we would when getting the receipt for a transaction that we submitted ourselves.

Reply:
no, the relayer doesn't wait at all, it just submits the tx to the mempool.

Reply:
I asked the question above, because I wanted to understand how we handle the scenario where the relayer submits the transaction, but sets the gas price so low that the transaction won't get picked up quickly, or might never get mined.

Reply:
right, in that case we can try all the relayers concurrently - I think we discussed this but didn't fully come to a conclusion? if the relayer has enough of a reward for the small amount of gas lost to potentially reverted txs (which they should), I think that's a decent solution

Reply:
That could work. There is still the question of the first attempt though, if we don't go to all of them immediately.

Our existing `WaitForReceipt` method inside `waitAndCheck` will block for 12 minutes at minimum before failing. It would be nice to fail much faster in 2 scenarios:
* Relayer gives us a made up, random transaction ID
* Relayer set the gas price so low that the transaction won't get mined (or won't get mined anytime soon)

Reply:
@dimalinux I've updated the `waitAndCheck` function so it now checks to see if the tx is in the mempool
- if it's not in the mempool in 6s, it errors
- if it's not included in a block in 1min, it errors
let me know what you think

Reply:
Looks good.  I added some suggestions on how we might improve the code and a question on how we know that the transaction the relayer gave us is for our claim.